src/mclient/mcursor.c

```diff
@@ -130,75 +131,101 @@ static void select_image_events(Display *xdpy) {
-    Display *dpy = XOpenDisplay(NULL);
     if (!dpy) {
-        MLOGE("[ct] error calling XOpenDisplay\n");
-        return (void *)-1;
+    	MLOGE("Failed to open display.\n");
+    	return (void *)-1;
     }
 
-    /* check for XInputExtension (XI*) */
-    int xi_opcode, event, error;
-    if (!XQueryExtension(dpy, "XInputExtension",
-            &xi_opcode, &event, &error)) {
-        MLOGE("[ct] XInputExtension unavailable!\n");
-        XCloseDisplay(dpy);
-        return (void *)-1;
+    if (!XQueryExtension(dpy, "XInputExtension", &xi_opcode, &event, &error)) {
+       MLOGE("X Input extension not available.\n");
+       return (void *)-1;
     }
 
-    /* check for XI2 version */
-    int ret;
-    int major = XI_2_Major, minor = XI_2_Minor;
-    ret = XIQueryVersion(dpy, &major, &minor);
-    if (ret == BadRequest) {
-        MLOGE("[ct] No matching XI2 support. (%d.%d only)\n", major, minor);
-        XCloseDisplay(dpy);
-        return (void *)-1;
-    }
+    if (!has_xi2(dpy))
+       return (void *)-1;
 
-    /* Get the main pointer device id for XI2 requests */
-    int pointer_dev_id;
-    XIGetClientPointer(dpy, None, &pointer_dev_id);
+    /* select for XI2 events */
+    select_events(dpy, DefaultRootWindow(dpy));
 
-    XIEventMask evmask;
-    /*
-     * The mask is set here in a very non-intuitive way
-     * but is more robust for adding additional event masks.
-     *
-     * (1) check how many bytes you need with XIMaskLen and
-     *     alloc an array of unsigned chars accordingly.
-     */
-    unsigned char mask[XIMaskLen(XI_Motion)] = { 0 };
-    evmask.deviceid = pointer_dev_id;
-    evmask.mask_len = sizeof(mask);
-    evmask.mask = mask;
-    /*
-     * (2) use XISetMask macro to toggle the right event bits
-     */
-    XISetMask(mask, XI_Motion);
-    XISelectEvents(dpy, DefaultRootWindow(dpy), &evmask, 1);
+    while(1) 
+    {
+    	XGenericEventCookie *cookie = &ev.xcookie;
+    	Window      	root_ret, child_ret;
+    	int         	root_x, root_y;
+    	int         	win_x, win_y;
+    	unsigned int    mask;
 
-    XEvent ev;
-    do {
-        XNextEvent(dpy, &ev);
-        /* see http://who-t.blogspot.com/2009/07/xi2-and-xlib-cookies.html */
-        if (XGetEventData(dpy, &ev.xcookie)) {
-            XGenericEventCookie *cookie = &ev.xcookie;
-            if (cookie->extension == xi_opcode &&
-                cookie->evtype == XI_Motion) {
-                XIDeviceEvent *xiev = (XIDeviceEvent *)cookie->data;
-                update_cursor(dpy, this->mMdpy, &this->mBuffer,
-                    (int)xiev->root_x, (int)xiev->root_y);
-            }
+    	XNextEvent(dpy, &ev);
+
+    	if (cookie->type != GenericEvent ||cookie->extension != xi_opcode )
+           continue;
+
+        if (XGetEventData(dpy, cookie))
+        {
+    	    if (cookie->evtype == XI_RawMotion) 
+            {
+        	XQueryPointer(dpy, DefaultRootWindow(dpy),&root_ret, &child_ret, &root_x, &root_y, &win_x, &win_y, &mask);
```

Comment:
Interesting...so we actually need to query the mouse pointer location when triggered by XI_RawMotion instead of using the cookie data? I wonder why X did not report the locations correctly with XI_Motion. Seems like a possible bug in X itself: https://github.com/glfw/glfw/issues/304. I think this will add a little more latency since we make one more server call for each motion event but it's ok for now. I would like to find a way to use the `cookie->data` to get the cursor position if possible to avoid the second call.